_posts/2019-11-15-enforcing-database-transactions-with-rust.md

```diff
@@ -0,0 +1,146 @@
+---
+layout: post
+title: Enforcing database transactions with Rust
+author: jgke
+excerpt: >
+  Rust's ownership model enables enforcing the use of database transactions,
+  while still allowing single queries without transactions.
+tags:
+ - Database
+ - Rust
+---
+
+Consider this Clojure snippet:
+
+```clojure
+(defn fetch-thing [user-id thing-id]
+  (if (db/has-permission user-id thing-id)
+    {:status 200 :body (db/fetch-thing thing-id)}
+    {:status 400})
+```
+
+It's clean, does clearly what it looks like it does, but there's a small
+problem; it doesn't use a transaction. If something happens in the database
+between the second and the third line, for example if the `thing` referenced by
+`thing-id` is deleted, the output of `db/fetch-thing` can get quite
+unpredictable. Could we make this kind of programs safer?
+
+## Enter Rust
+
+Rust is a relatively new programming language, with focus on speed and safety.
+Rust has the unique notion of 'ownership', which enables memory safety without
+relying on a garbage collector. Nowadays in most languages all variables are
+actually references, which are just thrown around and are either wildly mutable
+or completely immutable. Rust provides a way to write code where mutation is
+explicit and safe.
+
+In my opinion, the best benefit of the borrow checker is not the speedup it
```

Comment:
Jumps in to talking about the benefits of the borrow checker without actually introducing it. Maybe add a few lines on how it relates to the concepts in the previous paragraph?